torch/nn/parallel/replicate.py

```diff
@@ -83,26 +106,124 @@ def _copy_scriptmodule_methods(modules, module_copies, module_indices):
             replica._copy_method(method_name, param_list, module)
 
 
+# group tensors on the same device together, which can later be broadcast to
+# a list of devices. For example,consider 5 tensors on 2 devices
+#   a = torch.Tensor(0).cuda(0)
+#   b = torch.Tensor(0).cuda(0)
+#   c = torch.Tensor(0).cuda(1)
+#   d = torch.Tensor(0).cuda(0)
+#   e = torch.Tensor(0).cuda(1).
+# Let inputs be
+#   tensors = [a, b, c, d, e] and
+#   devices = [[0, 1], [2, 3]].
+# Then, outputs will be:
+#   grouped_tensors = [[a, b, d], [c, e]],
+#   grouped_devices = [[0, 2], [1, 3]],
+#   original_index = [[0, 1, 3], [2, 4]],
+# meaning that grouped_tensors[i] will be broadcast to grouped_devices[i].
+def _group_by_device(tensors, devices):
+    if isinstance(devices[0], list):
+        # device id to output group index, this is necessary when `tensors` only
+        # use a subset of devices in `devices[0]`
+        dev_to_idx = {dev: i for i, dev in enumerate(
+            set([t.device.index for t in tensors]))}
+
+        # Group tensors by devices and remember each tensor's original index.
+        # The original_index helps to recover the original input tensor order
+        # from grouped tensors.
+        grouped_tensors = [[] for _ in range(len(dev_to_idx))]
+        original_index = [[] for _ in range(len(dev_to_idx))]
+        for i, t in enumerate(tensors):
+            group_id = dev_to_idx[t.device.index]
+            original_index[group_id].append(i)
+            grouped_tensors[group_id].append(t)
+
+        # group devices together if they should be in the same broadcast call
+        grouped_devices = [[] for _ in range(len(dev_to_idx))]
+        transpose = list(zip(*devices))
+        for row in transpose:
+            if row[0] in dev_to_idx:
+                grouped_devices[dev_to_idx[row[0]]] = list(row)
+
+        return grouped_tensors, grouped_devices, original_index
+    else:
+        return [tensors], [devices], [list(range(len(tensors)))]
+
+
+# Return len(devices) replicas of input tensors. If input tensors reside on
+# multiple GPUs, devices must be a 2D list with devices[0] matching input
+# tensors' devices.
```

Comment:
Can you include a sample return value, like in `_group_by_device`?

That would be super helpful to figure out what the second half of this function is doing. It's a bit hard to figure out just by looking at the code and reverse engineering the return type/semantics.

Reply:
yes, will do.